README.md
# openwrt_ap502

ChangeLog
v1.0 2017-10-19


src/iw.c
/*
 * nl80211 userspace tool
 *
 * Copyright 2007, 2008	Johannes Berg <johannes@sipsolutions.net>
 */

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <net/if.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdbool.h>
#include <net/if.h>
#include <netinet/in.h>
#include <linux/if_slip.h>
#include <sys/ioctl.h>
#include <syslog.h>

#include <netlink/genl/genl.h>
#include <netlink/genl/family.h>
#include <netlink/genl/ctrl.h>
#include <netlink/msg.h>
#include <netlink/attr.h>

#include "nl80211.h"
#include "iw.h"
#include "debug.h"

extern int default_level;

/* libnl 1.x compatibility code */
#if !defined(CONFIG_LIBNL20) && !defined(CONFIG_LIBNL30)
static inline struct nl_handle *nl_socket_alloc(void)
{
	return nl_handle_alloc();
}

static inline void nl_socket_free(struct nl_sock *h)
{
	nl_handle_destroy(h);
}

static inline int nl_socket_set_buffer_size(struct nl_sock *sk,
					    int rxbuf, int txbuf)
{
	return nl_set_buffer_size(sk, rxbuf, txbuf);
}
#endif /* CONFIG_LIBNL20 && CONFIG_LIBNL30 */

int iw_debug = 0;

static int error_handler(struct sockaddr_nl *nla, struct nlmsgerr *err,
			 void *arg)
{
	int *ret = arg;
	*ret = err->error;
	return NL_STOP;
}

static int finish_handler(struct nl_msg *msg, void *arg)
{
	int *ret = arg;
	*ret = 0;
	return NL_SKIP;
}

static int ack_handler(struct nl_msg *msg, void *arg)
{
	int *ret = arg;
	*ret = 0;
	return NL_STOP;
}

int nl80211_init(struct nl80211_state *state)
{
	int err;

	state->nl_sock = nl_socket_alloc();
	if (!state->nl_sock) {
		debug(LOG_ERR, "Failed to allocate netlink socket.\n");
		return -ENOMEM;
	}

	nl_socket_set_buffer_size(state->nl_sock, 8192, 8192);

	if (genl_connect(state->nl_sock)) {
		debug(LOG_ERR, "Failed to connect to generic netlink.\n");
		err = -ENOLINK;
		nl_socket_free(state->nl_sock);
		return err;
	}

	state->nl80211_id = genl_ctrl_resolve(state->nl_sock, "nl80211");
	if (state->nl80211_id < 0) {
		debug(LOG_ERR, "nl80211 not found.\n");
		err = -ENOENT;
		nl_socket_free(state->nl_sock);
		return err;
	}
	state->msg = nlmsg_alloc();
	if (!state->msg) {
		debug(LOG_ERR, "failed to allocate netlink message\n");
		return 2;
	}

	if (default_level > LOG_WARNING){
		state->cb = nl_cb_alloc(NL_CB_DEBUG);
		state->s_cb = nl_cb_alloc(NL_CB_DEBUG);
	} 
	else {
		state->cb = nl_cb_alloc(NL_CB_DEFAULT);
		state->s_cb = nl_cb_alloc(NL_CB_DEFAULT);
	}
	if (!state->cb || !state->s_cb) {
		debug(LOG_ERR, "failed to allocate netlink callbacks\n");
		goto err;
	}

	nl_socket_set_cb(state->nl_sock, state->s_cb);

	nl_cb_err(state->cb, NL_CB_CUSTOM, error_handler, &state->err);
	nl_cb_set(state->cb, NL_CB_FINISH, NL_CB_CUSTOM, finish_handler, &state->err);
	nl_cb_set(state->cb, NL_CB_ACK, NL_CB_CUSTOM, ack_handler, &state->err);

	return 0;

err:
	return 1;
}

void nl80211_cleanup(struct nl80211_state *state)
{
	nl_cb_put(state->cb);
	nlmsg_free(state->msg);
	nl_socket_free(state->nl_sock);
}

static int phy_lookup(char *name)
{
	char buf[200];
	int fd, pos;
	
	debug(LOG_DEBUG, "phy:%s\n", name);

	snprintf(buf, sizeof(buf), "/sys/class/ieee80211/%s/index", name);

	fd = open(buf, O_RDONLY);
	if (fd < 0){
		debug(LOG_ERR, "open %s fail <%s>\n", buf, strerror(errno));
		return -1;
	}
	pos = read(fd, buf, sizeof(buf) - 1);
	if (pos < 0) {
		close(fd);
		return -1;
	}
	buf[pos] = '\0';
	close(fd);
	return atoi(buf);
}

int ieee80211_channel_to_frequency(int chan, enum nl80211_band band)
{
	/* see 802.11 17.3.8.3.2 and Annex J
	 * there are overlapping channel numbers in 5GHz and 2GHz bands */
	if (chan <= 0)
		return 0; /* not supported */
	switch (band) {
		case NL80211_BAND_2GHZ:
			if (chan == 14) 
				return 2484;
			else if (chan < 14) 
				return 2407 + chan * 5;
			break;
		case NL80211_BAND_5GHZ:
			if (chan >= 182 && chan <= 196)
				return 4000 + chan * 5;
			else
				return 5000 + chan * 5;
			break;
		case NL80211_BAND_60GHZ:
			if (chan < 5)
				return 56160 + chan * 2160;
			break;
		default:
			;   
	}   
	return 0; /* not supported */
}

int interface_up(char * name)
{
	struct ifreq ifr;
	int sockfd;
	int ret = 0;

	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
	if (sockfd < 0){
		debug(LOG_ERR, "create socket error:%s\n", strerror(errno));	
		return -1;
	}
	memset(&ifr, 0, sizeof(struct ifreq));  
	strncpy(ifr.ifr_name, name, IFNAMSIZ);
	ret = ioctl(sockfd, SIOCGIFFLAGS, &ifr);
	if (ret < 0){
		debug(LOG_ERR, "ioctl SIOCGIFFLAGS error:%s\n", strerror(errno));
		goto out;
	}
	ifr.ifr_flags |= (IFF_UP | IFF_RUNNING);
	ret = ioctl(sockfd, SIOCSIFFLAGS, &ifr);
	if (ret < 0){
		debug(LOG_ERR, "ioctl SIOCGIFFLAGS error:%s\n", strerror(errno));
		goto out;
	}

out:
	close(sockfd);
	return ret;
}

int interface_channel(struct nl80211_state *state)
{
	unsigned int freq = state->channel;
	enum nl80211_band band;
	state->devidx = phy_lookup("phy0");
	if (state->devidx < 0){
		debug(LOG_ERR, "devidx:%lld\n", state->devidx);
		return -errno;
	}
	nlmsg_free(state->msg);
	state->msg = nlmsg_alloc();
	if (!state->msg) {
		debug(LOG_ERR, "failed to allocate netlink message\n");
		return 2;
	}

	band = freq <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
	freq = ieee80211_channel_to_frequency(freq, band);
	printf("freq :%d\n", freq);
	genlmsg_put(state->msg, 0, 0, state->nl80211_id, 0, 0,NL80211_CMD_SET_WIPHY, 0);
	NLA_PUT_U32(state->msg, NL80211_ATTR_WIPHY, state->devidx);
	NLA_PUT_U32(state->msg, NL80211_ATTR_WIPHY_FREQ, freq);
	NLA_PUT_U32(state->msg, NL80211_ATTR_WIPHY_CHANNEL_TYPE, NL80211_CHAN_NO_HT);
	state->err = nl_send_auto_complete(state->nl_sock, state->msg);
	if(state->err < 0){
		nl_perror(state->err, "nl_send_auto_complete");
		return state->err;
	}

	state->err = 1;
	while (state->err > 0)
		nl_recvmsgs(state->nl_sock, state->cb);

	return state->err;	
nla_put_failure: /* jump from NLA_PUT_U32  */
	state->err = -ENOBUFS;
	return state->err;
}

int interface_add(struct nl80211_state *state)
{
	state->devidx = phy_lookup("phy0");
	if (state->devidx < 0){
		debug(LOG_ERR, "devidx:%lld\n", state->devidx);
		return -errno;
	}
	genlmsg_put(state->msg, NL_AUTO_PID, NL_AUTO_SEQ, state->nl80211_id, 0, 0, NL80211_CMD_NEW_INTERFACE, 0);
	NLA_PUT_U32(state->msg, NL80211_ATTR_WIPHY, state->devidx);
	NLA_PUT_STRING(state->msg, NL80211_ATTR_IFNAME, state->ifc);
	NLA_PUT_U32(state->msg, NL80211_ATTR_IFTYPE, NL80211_IFTYPE_MONITOR);
	state->err = nl_send_auto_complete(state->nl_sock, state->msg);
	if(state->err < 0){
		nl_perror(state->err, "nl_send_auto_complete");
		return state->err;
	}
	state->err = 1;
	while (state->err > 0)
		nl_recvmsgs(state->nl_sock, state->cb);

	return state->err;	
nla_put_failure:
	state->err = -ENOBUFS;
	return state->err;

}

int interface_del(struct nl80211_state *state)
{
	state->devidx = if_nametoindex(state->ifc);
	if (state->devidx < 0){
		debug(LOG_ERR, "devidx:%lld\n", state->devidx);
		return -errno;
	}
	nlmsg_free(state->msg);
	state->msg = nlmsg_alloc();
	if (!state->msg) {
		debug(LOG_ERR, "failed to allocate netlink message\n");
		return 2;
	}
	genlmsg_put(state->msg, NL_AUTO_PID, NL_AUTO_SEQ, state->nl80211_id, 0, 0, NL80211_CMD_DEL_INTERFACE, 0);
	NLA_PUT_U32(state->msg, NL80211_ATTR_IFINDEX, state->devidx);
	state->err = nl_send_auto_complete(state->nl_sock, state->msg);
	debug(LOG_DEBUG, "delete interface:%s\n", state->ifc);
	return state->err;	
nla_put_failure:
	state->err = -ENOBUFS;
	return state->err;
}

#if 0
int main(int argc, char **argv)
{
	struct nl80211_state nlstate;
	int err;

	/* calculate command size including padding */

	err = nl80211_init(&nlstate);
	if (err)
		return 1;
	argv++;
	err = interface_add(&nlstate, "test");
	printf("err:%d\n", err);
	if (err < 0) {
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
	} 

	err = interface_up("test");
	if (err < 0){
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
	}

	err = interface_channel(&nlstate, 11);
	if (err < 0){
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
	}

#if 0
	err = interface_del(&nlstate, "test");
	if (err < 0){
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
	}
#endif

	nl80211_cleanup(&nlstate);

	return err;
}
#endif 


src/iw.h
#ifndef __IW_H
#define __IW_H

#include <stdbool.h>
//#include <netlink/netlink.h>
#include <netlink/genl/genl.h>
#include <netlink/genl/family.h>
#include <netlink/genl/ctrl.h>
#include <endian.h>

#include "nl80211.h"
#include "ieee80211.h"

#define ETH_ALEN 6

/* libnl 1.x compatibility code */
#if !defined(CONFIG_LIBNL20) && !defined(CONFIG_LIBNL30)
#  define nl_sock nl_handle
#endif

struct nl80211_state {
	struct nl_msg *msg;
	struct nl_cb *cb;
	struct nl_cb *s_cb;
	struct nl_sock *nl_sock;
	int32_t nl80211_id;
	int64_t devidx;
	uint8_t *ifc;
	uint32_t channel;
	int32_t err;
};

enum command_identify_by {
	CIB_NONE,
	CIB_PHY,
	CIB_NETDEV,
	CIB_WDEV,
};

enum id_input {
	II_NONE,
	II_NETDEV,
	II_PHY_NAME,
	II_PHY_IDX,
	II_WDEV,
};

int nl80211_init(struct nl80211_state *state);
void nl80211_cleanup(struct nl80211_state *state);
int ieee80211_channel_to_frequency(int chan, enum nl80211_band band);
int interface_channel(struct nl80211_state *state);
int interface_up(char * name);
int interface_add(struct nl80211_state *state);
int interface_del(struct nl80211_state *state);


#endif /* __IW_H */


src/debug.c
#include <stdio.h>
#include <errno.h>
#include <syslog.h>
#include <stdarg.h>
#include <time.h>
#include <unistd.h>
#include <stdlib.h>
#include <libgen.h>
#include <string.h>
#include <arpa/inet.h>

#include "debug.h"

typedef unsigned char uint8_t;

extern int default_level;
/** @internal
  Do not use directly, use the debug macro */
	void
_debug(const char *filename, int line, int level, const char *format, ...)
{
	char buf[28];
	va_list vlist;
	time_t ts;

	time(&ts);

	if (default_level >= level) {
		if (level <= LOG_WARNING) {
			fprintf(stderr, "[%d][%.24s][%u](%s:%d) ", level, ctime_r(&ts, buf), getpid(),
					filename, line);
			va_start(vlist, format);
			vfprintf(stderr, format, vlist);
			va_end(vlist);
			fputc('\n', stderr);
		} else if (DAEMON) {
			fprintf(stdout, "[%d][%.24s][%u](%s:%d) ", level, ctime_r(&ts, buf), getpid(),
					filename, line);
			va_start(vlist, format);
			vfprintf(stdout, format, vlist);
			va_end(vlist);
			fputc('\n', stdout);
			fflush(stdout);
		}

		if (LOG_TO_SYSLOG) {
			openlog("WDS-SON", LOG_PID, LOG_DAEMON);
			va_start(vlist, format);
			vsyslog(level, format, vlist);
			va_end(vlist);
			closelog();
		}
	}
}

void hexdump(const uint8_t *data, int32_t len)
{
	int32_t i;
	uint8_t *temp = (uint8_t *)malloc(len * 6);
		
	if(LOG_TO_SYSLOG){
		for(i = 0; i < len; i++){
			snprintf(temp + i * 3,  len * 6 - i * 3, "%02x ", data[i]);
		}
		openlog("WDS-SON", LOG_PID, LOG_DAEMON);
		syslog(LOG_INFO, "hexdump length: %d", len);
		syslog(LOG_INFO, "%s", temp);
		closelog();
		free(temp);
	}
}



src/beacon.h
#ifndef _BEACON_H__
#define _BEACON_H__

#define BEACON_INTERVAL 10240
#define MAC_ADDR_LEN 17 
#define INTERFACE_LEN 32
#define ESSID_LEN 32
#define PASSWD_LEN 63
#define BEACON_ELEMENT_ID 0

#define ELEMENT_BSSID 200
#define ELEMENT_SSID 201
#define ELEMENT_KEY  202

#define FAKEAP "cellwifi-fakeap"
#define FAKEAP_LEN 15

#define CAPTURE_PID_FILE "/var/run/capture.pid"
#define BEACON_PID_FILE "/var/run/beacon.pid"

#define CRYPTO_KEY_LEN 32
#define CRYPTO_KEY "zhongyicaiwdskey"
#define INIT_VECT "1234567890abcdef"

#define WDS_SCRIPT "/usr/bin/wds-son "

#define PCAP_FILTER "subtype beacon and ether src 00:00:00:00:00:00"

struct FrameControl {
	uint8_t version:2;
	uint8_t type:2;
	uint8_t subtype:4;
} __attribute__((packed));

struct element {
	uint8_t elementID;
	uint8_t length;
	uint8_t *data;
} __attribute__((packed));

struct mgt_beacon {
	uint64_t timestamp;
	uint16_t interval;
	uint16_t cap;
} __attribute__((packed));


struct ieee80211_radiotap_header {
	uint8_t        it_version;     /* set to 0 */
	uint8_t        it_pad;
	uint16_t       it_len;         /* entire length */
	uint32_t       it_present;     /* fields present */
} __attribute__((packed));

struct rtapdata {
	uint8_t  flags;
	uint8_t  datarate;
	uint16_t frequency;
	uint16_t  type;
	int8_t  ssi;        
	uint8_t antenna;
	uint16_t rx_flags;
} __attribute__ ((packed));

struct AP_descriptor {
	uint8_t MAC[MAC_ADDR_LEN];
	uint8_t channel;
	uint8_t ssid_len;
	int8_t ssid[ESSID_LEN];
	uint8_t key_len;
	int8_t key[PASSWD_LEN];
	uint8_t interface[INTERFACE_LEN];
};

struct MAC_HEADER {
	uint16_t framecontrol;
	uint16_t duration;
	uint8_t dest[6];
	uint8_t src[6];
	uint8_t bssid[6];
	uint16_t sequence;
	struct mgt_beacon beacon;
} __attribute__((packed));

const uint16_t CHANNEL[11] = {
	0x6C09, 0x7109, 0x7609, 0X7B09,
	0x8009, 0x8509, 0x8A09, 0x8F09,
	0x9409, 0x9909, 0x9E09
};

#endif


src/beacon.c
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <netpacket/packet.h>
#include <linux/if_ether.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <openssl/aes.h>
#include <regex.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <syslog.h>
#include <signal.h>

//#define _BSD_SOURCE             /* See feature_test_macros(7) */
#include <endian.h>

#include "beacon.h"
#include "debug.h"
#include "iw.h"

struct ieee80211_radiotap_header rthdr;
struct rtapdata rtp;
struct MAC_HEADER header; 
struct AP_descriptor ap;
int32_t default_level = LOG_WARNING;
static struct nl80211_state nlstate;
int32_t sock = -1;

void usage(void)
{
	printf("\n******************************************\n");
	printf("beacon <arg>\n");
	printf("-i : interface to send beacon frame\n");
	printf("-k : password \n");
	printf("-s : ssid\n");
	printf("-m : mac\n");
	printf("-d : debug level\n");
	printf("-h : show help message\n");
	printf("\n******************************************\n");
}

int32_t isValidMac(const int8_t *mac)
{
	int8_t *reg="^([0-9a-fA-F]{2})(([/\\s:][0-9a-fA-F]{2}){5})$";
	regex_t pat_cmdline;
	regmatch_t matches[17];
	if(regcomp(&pat_cmdline, reg, REG_EXTENDED)){
		debug(LOG_ERR, "regcomp error <%s>\n", strerror(errno));	
		exit(1);
	}
	if(regexec(&pat_cmdline, mac, 17, matches, 0)){
		regfree(&pat_cmdline);	
		return -1;
	}
	regfree(&pat_cmdline);
	return 0;
}

int update_pid_file(void)
{
	char path[] = BEACON_PID_FILE;
	int fd = 0;  

	unlink(path);

	if ( (fd = open(path, O_CREAT|O_WRONLY|O_TRUNC, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH )) < 0 ) {  //check permissions of generated file
		debug(LOG_ERR, "could not open %s - %s", path, strerror(errno) );
		return -1; 
	}        

	dprintf(fd, "%d\n", getpid());
	close (fd);

	return 0;
}


void parse_arg(int argc, char **argv)
{
	int32_t c;
	memset(&ap, 0, sizeof(ap));
	while (EOF!= (c = getopt(argc, argv, "hi:k:m:s:d:"))) {
		switch(c){
			case 'i':{
				memcpy(ap.interface, optarg, 32);
				debug(LOG_DEBUG, "ap.interface:%s\n", ap.interface);
				break;
			}
			case 'k':{
				ap.key_len = strlen(optarg) > PASSWD_LEN ? PASSWD_LEN : strlen(optarg);
				memcpy(ap.key, optarg, ap.key_len);
				break;
			}
			case 's':{
				ap.ssid_len = strlen(optarg) > ESSID_LEN ? ESSID_LEN : strlen(optarg);
				memcpy(ap.ssid, optarg, ap.ssid_len);
				break;
			}
			case 'm':{
				if(isValidMac(optarg)){
					debug(LOG_ERR, "Invalid MAC\n");
					usage();
					exit(1);

				}
				debug(LOG_DEBUG, "mac:%s\n", optarg);
				memcpy(ap.MAC, optarg, MAC_ADDR_LEN);		 
				break;
			}
			case 'd':{
				default_level = atoi(optarg);
				break;
			}
			case 'h':
			case '?':
			default: {
				usage();
				exit(1);
			}
		}
		
	}

	debug(LOG_DEBUG, "ap.ssid_len:%d\n", ap.ssid_len);
	debug(LOG_DEBUG, "ap.ssid:%s\n", ap.ssid);
	debug(LOG_DEBUG, "ap.key_len:%d\n", ap.key_len);
	debug(LOG_DEBUG, "ap.key:%s\n", ap.key);

	if (!ap.ssid_len || !ap.key_len || !ap.ssid || !ap.key){
		usage();	
		exit(1);	
	}
	
}

void config_init(struct AP_descriptor *ap)
{
	rthdr.it_version = 0;
	rthdr.it_pad = 0;
	rthdr.it_len = htole16(18);
	/*	we use a more general present 	*/
	/*	see http://www.radiotheader.org/ for more introduction */
	rthdr.it_present =	0x2e480000;

	/* IEEE80211_RADIOTheader_F_FCS; */
	/* auto add frame check sequence */
	rtp.flags = 0;

	/* 1.0 Mb/s */
	rtp.datarate = 0x02;

	/* channel frequency , will be fixed by drivers*/
	rtp.frequency = 0x6C09;

	/* 802.11g */
	rtp.type = 0xC000;

	/* -20dbm, will be fixed by drivers */
	rtp.ssi = 0xA0;

	/* has antenna , will be fixed by drivers */
	rtp.antenna = 0x01;

	rtp.rx_flags = 0x0000;

	/* beacon frame */
	header.framecontrol = 0x8000;

	/* duration 0s */
	header.duration = 0x0000;

	/* destination MAC address, broadcast */ 
	memcpy(header.dest, "\xFF\xFF\xFF\xFF\xFF\xFF", 6);

	/* source MAC address, fake MAC address */
	memcpy(header.src, "\x00\x00\x00\x00\x00\x00", 6);

	/* BSSID , fake MAC address */
	memcpy(header.bssid, "\x00\x00\x00\x00\x00\x00", 6);
	
	/* fragments and sequence number: start from 1 */
	header.sequence = 0x0001;

	/* beacon timestamp for BSS : initial value to 0 */
	header.beacon.timestamp = 0LL;

	/* beacon interval : 10TU*/
	header.beacon.interval = htole16(BEACON_INTERVAL / 1024);

	/* beacon capabilities */
	header.beacon.cap = htole16(0x0001);

	//ap->ssid_len = 6;
	//memcpy(ap->ssid, "fakeap", 6);
	//ap->key_len = 10;
	//memcpy(ap->key, "1234567890", 10);
}

uint8_t *cryptoBeacon(const uint8_t *in, uint8_t in_len, uint8_t *out)
{
	uint8_t iv[AES_BLOCK_SIZE * 4] = INIT_VECT;
	uint8_t cipher[AES_BLOCK_SIZE * 16]; 
	uint8_t key[CRYPTO_KEY_LEN] = CRYPTO_KEY;
	int32_t nr_of_bits = 0;
	int32_t nr_of_bytes = 0;
	AES_KEY a_key;
	
	memset(cipher, 0, sizeof(cipher));
	nr_of_bits = 8 * sizeof(key);
	nr_of_bytes = in_len; 
	AES_set_encrypt_key(key, nr_of_bits, &a_key);
	AES_cbc_encrypt(in, cipher, nr_of_bytes, &a_key, iv, AES_ENCRYPT);
	int i = 0;
	memcpy(out, cipher, in_len);
	return out;
}

uint16_t constructBeacon(struct AP_descriptor *ap, uint8_t *data)
{
	/****************************************************************/
	/*					Big endian									*/
	/*	 Frame Control		0×80 0×00								*/
	/*	 Duration	0×00 0×00										*/
	/*	 Destination Address FF:FF:FF:FF:FF:FF						*/
	/*	 Source Address		 00:11:22:33:44:55						*/
	/*	 BSSID				 00:11:22:33:44:55						*/
	/*	 Seq-ID	12-bit Sequence Number + 4-bit Fragment Number		*/
	/*	 Timestamp	BSS timestamp	8-bypes							*/
	/*	 Beacon interval	0×64 0×00  100TU						*/
	/*	 Capability info	0×01 0×00  PCF no poll					*/
	/*	 SSID	Element ID + len + data								*/
	/*	 FCS	auto-calcuate by drivers							*/
	uint16_t offset = 0;	
	uint8_t *pos = data;
	uint8_t cipher[AES_BLOCK_SIZE * 16];

	memcpy(pos, &rthdr, sizeof(struct ieee80211_radiotap_header));
	pos += sizeof(struct ieee80211_radiotap_header);
	memcpy(pos, &rtp, sizeof(struct rtapdata));
	pos += sizeof(struct rtapdata);

	memcpy(pos, &header, sizeof(struct MAC_HEADER));
	pos += sizeof(struct MAC_HEADER);
	/* fake ap ssid */
	*pos = BEACON_ELEMENT_ID;
	pos++;
	*pos = FAKEAP_LEN;
	pos++;
	memcpy(pos, FAKEAP, FAKEAP_LEN);
	pos += FAKEAP_LEN;

	/* bssid */
	*pos = ELEMENT_BSSID;
	pos++;
	*pos = MAC_ADDR_LEN;
	pos++;
	memcpy(pos, ap->MAC, MAC_ADDR_LEN);
	pos += MAC_ADDR_LEN;

	/* ssid */
	//pos++;
	*pos = ELEMENT_SSID;
	pos++;
	*pos = ap->ssid_len;
	pos++;
	memcpy(pos, ap->ssid, ap->ssid_len);
	
	/* vendor */
	pos += ap->ssid_len;
	*pos = ELEMENT_KEY;
	pos++;
//	*pos = ap->key_len;
	*pos = (ap->key_len / 16 + 1) * 16;
	pos++;
	memcpy(pos, cryptoBeacon(ap->key, (ap->key_len / 16 + 1) * 16, cipher), (ap->key_len / 16 + 1) * 16);
	pos += (ap->key_len / 16 + 1) * 16;

	return pos - data;
}


ssize_t transmitBeacon(struct AP_descriptor *ap, int32_t sock,  uint8_t *data, uint16_t length)
{
	ssize_t ret = 0;
	
	hexdump(data, length);

	ret = write(sock, data, length);
	if (ret != length){
		debug(LOG_ERR, "send beacon frame failed:%s\n", strerror(errno));
	}
	return ret;
}

int32_t open_raw_socket(const int8_t *iface)
{
	int32_t sock;
	struct ifreq ifr;
	struct sockaddr_ll ll;

	debug(LOG_DEBUG, "open interface <%s>\n", iface);

	sock = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
	if (sock < 0){
		debug(LOG_ERR, "create raw socket error: %s\n", strerror(errno));
		exit(1);
	}

	memset(&ifr, 0, sizeof(ifr));
	strncpy(ifr.ifr_name, iface, sizeof(ifr.ifr_name) - 1);
	if(ioctl(sock, SIOCGIFINDEX, &ifr) < 0){
		debug(LOG_ERR, "set socket <SIOCGIFINDEX> error: %s\n", strerror(errno));
		return -1;
	}

	memset(&ll, 0, sizeof(ll));
	ll.sll_family = AF_PACKET;
	ll.sll_ifindex = ifr.ifr_ifindex;
	ll.sll_protocol = htons(ETH_P_ALL);
	if(bind(sock, (struct sockaddr *)&ll, sizeof(ll)) < 0){
		debug(LOG_ERR, "bind socket <AP_PACKET> error: %s\n", strerror(errno));
		close(sock);
		return -1;
	}
	
	struct packet_mreq mr;
	memset(&mr, 0, sizeof(mr));
	mr.mr_ifindex = ll.sll_ifindex;
	mr.mr_type = PACKET_MR_PROMISC;
	if(setsockopt(sock, SOL_PACKET, PACKET_ADD_MEMBERSHIP, &mr, sizeof(mr)) < 0){
		debug(LOG_ERR, "set socket option <PACKET_MR_PROMISC> error: %s\n", strerror(errno));
		close(sock);
		return -1;
	}
	return sock;
}

void interface_exit(struct nl80211_state *state)
{
	uint32_t err;
	err = interface_del(state);
	if (err < 0){
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
	}

	nl80211_cleanup(state);
}

int32_t interface_init(struct nl80211_state *state)
{
	int err;

	/* calculate command size including padding */

	err = nl80211_init(state);
	if (err){
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
		exit(1);
	}

	err = interface_add(state);
	if (err < 0) {
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
		nl80211_cleanup(state);
		exit(1);
	} 

	err = interface_up(state->ifc);
	if (err < 0){
		debug(LOG_ERR, "command failed: %s (%d)\n", strerror(-err), err); //ENFILE
		interface_exit(state);
		exit(1);
	}

	return err;
}


void signal_handler(int32_t sig)
{	
	interface_exit(&nlstate);
	exit(0);
}

#if 0
void signal_pipe(int32_t sig)
{
	interface_exit(&nlstate);
	interface_init(&nlstate);

	if(sock > 0){
		close(sock);
	}

	sock = open_raw_socket(nlstate.ifc);
	if (sock < 0){
		debug(LOG_ERR, "create raw socket[%s] error <%s>\n", nlstate.ifc, strerror(errno));
		return 1;
	}
	
}
#endif

static void signal_init(void)
{
	struct sigaction sa;
	sa.sa_handler = signal_handler;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = SA_RESTART;

	/* Trap SIGTERM */
	if (sigaction(SIGTERM, &sa, NULL) == -1) {
		debug(LOG_ERR, "sigaction(): %s", strerror(errno));	
		exit(1);
	}

	/* Trap SIGQUIT */
	if (sigaction(SIGQUIT, &sa, NULL) == -1) {
		debug(LOG_ERR, "sigaction(): %s", strerror(errno));
		exit(1);
	}

	/* Trap SIGINT */
	if (sigaction(SIGINT, &sa, NULL) == -1) {
		debug(LOG_ERR, "sigaction(): %s", strerror(errno));
		exit(1);
	}
	
#if 0
	/* SIGPIPE */
	sa.sa_handler = signal_pipe;
	if (sigaction(SIGPIPE, &sa, NULL) == -1) {
		debug(LOG_ERR, "sigaction(): %s", strerror(errno));
		exit(1);
	}
#endif
}

int32_t reinit(void)
{
	interface_exit(&nlstate); 
	interface_init(&nlstate);

	if (sock > 0){
		close(sock);
	}
	sock = open_raw_socket(nlstate.ifc);
	if (sock < 0){
		debug(LOG_ERR, "create raw socket[%s] error <%s>\n", nlstate.ifc, strerror(errno));
		return 1;
	}

	return 0;
}

int main(int argc, char **argv)
{
	uint8_t buffer[4096];	
	uint16_t len;
	ssize_t ret;

	if(update_pid_file()){
		debug(LOG_ERR, "write beacon pid to file error\n");
		return -1;
	}

	parse_arg(argc, argv);
	config_init(&ap);
	if (ap.interface){
		nlstate.ifc = ap.interface;
	}
	else {
		nlstate.ifc = "monitor";
	}
	
	signal_init();
	interface_init(&nlstate);

	sock = open_raw_socket(nlstate.ifc);
	if (sock < 0){
		debug(LOG_ERR, "create raw socket[%s] error <%s>\n", nlstate.ifc, strerror(errno));
		return 1;
	}

	while(1) {
		len = constructBeacon(&ap, buffer);
		debug(LOG_DEBUG, "beacon length :%d\n", len);
		ret = transmitBeacon(&ap, sock, buffer, len);	
		if (ret < 0 && (errno == ENXIO || errno == ENETDOWN)){
			if (reinit()) {
				break;
			}
		}
		sleep(1);
	}
	
	interface_exit(&nlstate);

	return 0;
}



